File: NinePSharp/messages/Tflush.cs
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

namespace NinePSharp.Messages;

// size[4] Tflush tag[2] oldtag[2]
public readonly struct Tflush : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Tflush;
    public ushort Tag { get; }

    public ushort OldTag { get; }

    public Tflush(ushort tag, ushort oldtag)
    {
        Tag = tag;
        OldTag = oldtag;
        Size = NinePConstants.HeaderSize + 2;
    }

    public Tflush(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));

        int offset = NinePConstants.HeaderSize;
        OldTag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(offset, 2));
    }

    public void WriteTo(Span<byte> data)
    {
        data.WriteHeaders(Size, Tag, Type);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt16LittleEndian(data.Slice(offset, 2), OldTag);
    }
}

File: NinePSharp.Messages.FSharp/messages/Tflush.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Tflush =
    { Size: uint32
      Tag: uint16
      OldTag: uint16 }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Tflush
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Tflush span &offset
            BinaryPrimitives.WriteUInt16LittleEndian(span.Slice(offset, 2), this.OldTag)

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let oldtag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(offset, 2))
        { Size = size; Tag = tag; OldTag = oldtag }

